src/features/orders/routes/add.order.route.ts

```diff
@@ -0,0 +1,219 @@
+
+          if (!Array.isArray) {
+            throw new Error('Invalid input format - expected array');
+          }
+          if (Array.isArray(ordersToStore)) {
+            //2. check if products exist and there's enough quantity
+            const productsInOrder = ordersToStore.reduce(
+              (
+                products: { _id: ObjectId; item: any; store: string }[],
+                order: any
+              ) => {
+                const itemsArray = order.items.reduce(
+                  (
+                    acc: { _id: ObjectId; item: any; store: ObjectId }[],
+                    element: any
+                  ) => {
+                    acc.push({
+                      _id: new ObjectId(element._id),
+                      item: getOrderItem(element),
+                      store: order.store,
+                    });
+                    return acc;
+                  },
+                  []
+                );
+
+                itemsArray.forEach(
+                  (element: { _id: ObjectId; item: any; store: string }) => {
+                    products.push(element);
+                  }
+                );
+                return products;
+              },
+              []
+            );
+
+            const productPromises = productsInOrder.map(
+              async (element: { _id: ObjectId; item: any }) => {
+                const product = await fastify
+                  .db(process.env.DB_NAME as string)
+                  ?.collection(CollectionEnum.Products)
+                  .findOne({ _id: new ObjectId(element._id) });
+
+                if (!product) {
+                  throw new Error(
+                    `Product with _id ${element.item._id} not found`
+                  );
+                }
+                if (product?.availableQuantity < element.item.quantity) {
+                  throw new Error(
+                    `Not enough quantity of product with _id ${element._id}`
+                  );
+                }
+
+                return product;
+              }
+            );
+
+            await Promise.all(productPromises);
+
+            // 3. update products quantity
+            await fastify
+              .db(process.env.DB_NAME as string)
+              ?.collection(CollectionEnum.Products)
+              .bulkWrite(
+                productsInOrder.map((el: { _id: ObjectId; item: any }) => ({
+                  updateOne: {
+                    filter: { _id: el._id },
+                    update: { $inc: { availableQuantity: -el.item.quantity } },
+                  },
+                })),
+                { session }
+              );
+
+            // 3. create order and order items one by one per store
+            await Promise.all(
+              ordersToStore.map(async (order) => {
+                const itms = productsInOrder
+                  .filter((item) => item.store === order.store)
```

Comment:
You've already traversed these products. It should be possible to create and use a more "performant" data structure for accessing the products from each store instead of traversing all of them for each one